src/compute/src/extensions/operator.rs

```diff
@@ -0,0 +1,371 @@
+                    data.swap(&mut buffer);
+                    output.session(&time).give_container(&mut buffer);
+                }
+
+                // We don't want to block compaction.
+                let mut upper = Antichain::new();
+                trace.read_upper(&mut upper);
+                trace.set_logical_compaction(upper.borrow());
+                trace.set_physical_compaction(upper.borrow());
+
+                let (size, capacity, allocations) = logic(&trace);
+
+                let size = size.try_into().expect("must fit");
+                if size != old_size {
+                    logger.log(ComputeEvent::ArrangementHeapSize {
+                        operator,
+                        size: size - old_size,
+                    });
+                }
+
+                let capacity = capacity.try_into().expect("must fit");
+                if capacity != old_capacity {
+                    logger.log(ComputeEvent::ArrangementHeapCapacity {
+                        operator,
+                        capacity: capacity - old_capacity,
+                    });
+                }
+
+                let allocations = allocations.try_into().expect("must fit");
+                if allocations != old_allocations {
+                    logger.log(ComputeEvent::ArrangementHeapAllocations {
+                        operator,
+                        allocations: allocations - old_allocations,
+                    });
+                }
+
+                old_size = size;
+                old_capacity = capacity;
+                old_allocations = allocations;
+            }
+        });
+    Arranged {
+        trace: arranged.trace,
+        stream,
+    }
+}
+
+impl<G, K, V, T, R> ArrangementSize for Arranged<G, TraceAgent<RowSpine<K, V, T, R>>>
+where
+    G: Scope<Timestamp = T>,
+    G::Timestamp: Lattice + Ord,
+    K: Data + Columnation,
+    V: Data + Columnation,
+    T: Lattice + Timestamp,
+    R: Semigroup,
+{
+    fn log_arrangement_size(self) -> Self {
+        log_arrangement_size_inner(self, |trace| {
+            let (mut size, mut capacity, mut allocations) = (0, 0, 0);
+            let mut callback = |siz, cap| {
+                allocations += 1;
+                size += siz;
+                capacity += cap
+            };
+            trace.map_batches(|batch| {
+                batch.layer.keys.heap_size(&mut callback);
+                batch.layer.vals.keys.heap_size(&mut callback);
+                vec_size(&batch.layer.offs, &mut callback);
+                vec_size(&batch.layer.vals.offs, &mut callback);
+                vec_size(&batch.layer.vals.vals.vals, &mut callback);
+            });
+            (size, capacity, allocations)
+        })
+    }
+}
+
+impl<G, K, T, R> ArrangementSize for Arranged<G, TraceAgent<RowKeySpine<K, T, R>>>
+where
+    G: Scope<Timestamp = T>,
+    G::Timestamp: Lattice + Ord,
+    K: Data + Columnation,
+    T: Lattice + Timestamp,
+    R: Semigroup,
```

Comment:
This is somewhat incorrect: `R` can be a complex type, such as a reduction monoid, but we're not counting this correctly without a `Columnation` bound. I'm working on a fix.

Reply:
It turns out that Differential doesn't currently support region-allocated `R`. This is because `OrderedLeaf` has a parameter to specify the container type, but where it's used in `OrdValBatch`, it's not specified. This requires a change to Differential to fix, or us defining our own batch types.

Reply:
I looked into this a bit more. Differential keeps for every leaf a `Vec<(T, R)>` maintaining how often a value appears at a certain time. When advancing time, it does so by advancing the `T` in-place, which requires mutable access to the data. With `Vec`, this is safe, but switching to `TimelyStack` does not allow safe mutable access to its elements. Indeed, it would only be safe if `advance_by` would not cause a reallocation, which in general we cannot guarantee. This means we cannot have a bound of `Columnation` on `R`, making it impossible to region-allocate various monoids.

For details, see TimelyDataflow/differential-dataflow#391.